crates/iota-sdk-ffi/src/uniffi_helpers.rs

```diff
@@ -1,65 +1,69 @@
 // Copyright (c) 2025 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use iota_graphql_client::pagination::Page;
+use iota_graphql_client::query_types::{Base64, PageInfo};
 use serde_json::Value;
 
 use crate::types::{
     checkpoint::CheckpointSummary,
     coin::Coin,
-    graphql::{
-        DynamicFieldOutput, Epoch, PageInfo, TransactionDataEffects, TransactionEvent, Validator,
-    },
+    graphql::{DynamicFieldOutput, Epoch, TransactionDataEffects, TransactionEvent, Validator},
     object::{MovePackage, Object},
     transaction::{SignedTransaction, TransactionEffects},
 };
 
+/// Defines a Page<T> where T is a uniffi Record to avoid generics
 macro_rules! define_paged_record {
     ($id:ident, $typ:ty) => {
-        #[derive(uniffi::Object, derive_more::From)]
-        pub struct $id(Page<$typ>);
-
-        #[uniffi::export]
-        impl $id {
-            pub fn page_info(&self) -> PageInfo {
-                self.0.page_info().clone().into()
-            }
-
-            pub fn data(&self) -> Vec<$typ> {
-                self.0.data().to_vec()
-            }
+        /// A page of items returned by the GraphQL server.
+        #[derive(Debug, Clone, uniffi::Record)]
+        pub struct $id {
+            /// Information about the page, such as the cursor and whether there are
+            /// more pages.
+            pub page_info: PageInfo,
+            /// The data returned by the server.
+            pub data: Vec<$typ>,
+        }
 
-            pub fn is_empty(&self) -> bool {
-                self.0.is_empty()
+        impl From<iota_graphql_client::pagination::Page<$typ>> for $id {
+            fn from(value: iota_graphql_client::pagination::Page<$typ>) -> Self {
+                Self {
+                    page_info: value.page_info.into(),
+                    data: value.data,
+                }
             }
         }
     };
 }
 
 define_paged_record!(CheckpointSummaryPage, CheckpointSummary);
+define_paged_record!(DynamicFieldOutputPage, DynamicFieldOutput);
+define_paged_record!(ValidatorPage, Validator);
 
+/// Defines a Page<T> where T is a uniffi Object to avoid generics
 macro_rules! define_paged_object {
     ($id:ident, $typ:ty) => {
-        #[derive(uniffi::Object, derive_more::From)]
-        pub struct $id(Page<$typ>);
-
-        #[uniffi::export]
-        impl $id {
-            pub fn page_info(&self) -> PageInfo {
-                self.0.page_info().clone().into()
-            }
-
-            pub fn data(&self) -> Vec<std::sync::Arc<$typ>> {
```

Comment:
test.py needs to be updated now to call .data instead of .data()

Reply:
done!